File: cs185/StudentSelecter/Java/src/WavePlayer.java
import java.io.*;
import java.nio.*;
import javax.swing.event.*;
import net.java.games.joal.*;
import net.java.games.joal.util.*;

/**
 * Represents a wave file player.
 */
public final class WavePlayer
{
	/**
     * The list of listeners to the source state changed event.
     */
    protected EventListenerList listenerList = new EventListenerList();
    
    /**
     * The OpenAL audio content used for playback.
     */
    private AL context;

    /**
     * The OpenAL sound buffer identifier.
     */
    private int[] buffer = new int[1];

    /**
     * The OpenAL sound source identifier.
     */
    private int[] source = new int[1];
    
    /**
     * The source state listener thread.
     */
    private Thread thread;

    /**
     * Initializes a new instance of the {@link WavePlayer} class.
     * @param fileName The file to read audio data from.
     * @throws IOException An I/O exception occurs.
     * @throws OpenALException An OpenAL exception occurs.
     */
    public WavePlayer(String fileName) throws IOException, OpenALException
    {
    	this();
        this.load(fileName);
    }

    /**
     * Initializes a new instance of the {@link WavePlayer} class.
     * @param stream The stream to read audio data from.
     * @throws IOException An I/O exception occurs.
     * @throws OpenALException An OpenAL exception occurs.
     */
    public WavePlayer(InputStream stream) throws IOException, OpenALException
    {
    	this();
        this.load(stream);
    }

    /**
     * Initializes a new instance of the {@link WavePlayer} class.
     * @throws OpenALException An OpenAL exception occurs.
     */
    private WavePlayer() throws OpenALException
    {
        this.context = ALFactory.getAL();
        
        ALut.alutInit();
        
        this.context.alGenBuffers(1, this.buffer);
        this.context.alGenSources(1, this.source);
        
        this.thread = new Thread(new Runnable()
        {
			@Override
			public void run()
			{
				WavePlayer.this.sourceStateListen();
			}
		});
        
        this.thread.start();
    }

    /**
     * Gets the state of the audio source; whether it is playing, paused or stopped.
     */
    public ALSourceState getSourceState()
    {
        int[] state = new int[1];
        this.context.alGetSourcei(this.source[0], AL.AL_SOURCE_STATE, state);
        return ALSourceState.fromInt(state[0]);
    }
    
    /**
     * Reads an 8-bit ASCII string from a stream.
     * @param stream The stream to read from.
     * @param count The number of bytes (e.g. the number of characters or the length of the string) to read.
     * @return The string, constructed from the individual bytes.
     * @throws IOException An I/O error occurs.
     */
    public static String readByteString(InputStream stream, int count) throws IOException
    {
    	char[] bytes = new char[count];
    	for (int i = 0; i < count; i++)
    	{
    		bytes[i] = (char)stream.read();
    	}
    	
    	return new String(bytes);
    }
    
    /**
     * Loads a wave/riff audio file.
     * Note that we have to reverse bytes in here because Java's streams use big endian format, whereas the WAVE file format is in little endian.
     * @param stream The stream to read from.
     * @return The data of the wave file.
     * @throws IOException 
     * @exception IllegalArgumentException {@link stream} is null -or- the stream is not a wave file -or- the wave file is not supported.
     * @exception IOException An I/O error occurs.
     */
    public static WaveData loadWave(InputStream stream) throws IOException
    {
        if (stream == null)
        {
            throw new IllegalArgumentException("stream is null");
        }

        DataInputStream reader = new DataInputStream(stream);

        // RIFF header signature
        if (!WavePlayer.readByteString(stream, 4).equals("RIFF"))
        {
            throw new IllegalArgumentException("Specified stream is not a wave file.");
        }

        @SuppressWarnings("unused")
		int riff_chunck_size = Integer.reverseBytes(reader.readInt());

        if (!WavePlayer.readByteString(stream, 4).equals("WAVE"))
        {
            throw new IllegalArgumentException("Specified stream is not a wave file.");
        }

        // WAVE header format signature
        if (!WavePlayer.readByteString(stream, 4).equals("fmt "))
        {
            throw new IllegalArgumentException("Specified wave file is not supported.");
        }

        @SuppressWarnings("unused")
		int format_chunk_size = Integer.reverseBytes(reader.readInt());
        @SuppressWarnings("unused")
		int audio_format = Short.reverseBytes(reader.readShort());
        int num_channels = Short.reverseBytes(reader.readShort());
        int sample_rate = Integer.reverseBytes(reader.readInt());
        @SuppressWarnings("unused")
		int byte_rate = Integer.reverseBytes(reader.readInt());
        @SuppressWarnings("unused")
		int block_align = Short.reverseBytes(reader.readShort());
        int bits_per_sample = Short.reverseBytes(reader.readShort());

        if (!WavePlayer.readByteString(stream, 4).equals("data"))
        {
            throw new IllegalArgumentException("Specified wave file is not supported.");
        }

        @SuppressWarnings("unused")
		int data_chunk_size = Integer.reverseBytes(reader.readInt());

        byte[] data = new byte[reader.available()];
        if (data.length == reader.read(data))
        {
        	return new WaveData(data, num_channels, bits_per_sample, sample_rate);
        }
        else
        {
        	return null;
        }
    }

    /**
     * Gets the OpenAL format specifier for the sound format with the specified number of channels and bits.
     * @param channels The number of channels in the sound format.
     * @param bits The number of bits in the sound format.
     */
    public static int getSoundFormat(int channels, int bits)
    {
        switch (channels)
        {
            case 1:
                return bits == 8 ? AL.AL_FORMAT_MONO8 : AL.AL_FORMAT_MONO16;
            case 2:
                return bits == 8 ? AL.AL_FORMAT_STEREO8 : AL.AL_FORMAT_STEREO16;
            default:
                throw new IllegalArgumentException("The specified sound format is not supported.");
        }
    }

    /**
     * Loads a wave file into the player.
     * @param fileName The name of the file to load.
     * @throws IOException An I/O exception occurs.
     */
    public void load(String fileName) throws IOException
    {
        this.load(new FileInputStream(fileName));
    }

    /**
     * Loads a wave file into the player.
     * @param stream The stream to load from.
     * @throws IOException An I/O exception occurs.
     */
    public void load(InputStream stream) throws IOException
    {
        WaveData sound_data = WavePlayer.loadWave(stream);
        ByteBuffer buffer = BufferUtils.newByteBuffer(sound_data.getAudioData().length);
        buffer.put(sound_data.getAudioData());
        this.context.alBufferData(this.buffer[0], WavePlayer.getSoundFormat(sound_data.getChannels(), sound_data.getRate()), buffer, sound_data.getAudioData().length, sound_data.getRate());

        this.context.alSourcei(this.source[0], AL.AL_BUFFER, this.buffer[0]);
    }

    /**
     * Plays the loaded wave file.
     * @throws InterruptedException The thread is interrupted while sleeping.
     */
    public void play()
    {
        this.context.alSourcePlay(this.source[0]);
    }

    /**
     * Frees resources.
     */
    public void dispose()
    {
    	this.context.alDeleteBuffers(1, this.buffer);
    	this.context.alDeleteSources(1, this.source);
        ALut.alutExit();
    }
    
    public void addSourceStateChangedListener(SourceStateChangedListener listener)
    {
    	this.listenerList.add(SourceStateChangedListener.class, listener); 
    }
    
    public void removeSourceStateChangedListener(SourceStateChangedListener listener)
    {
    	this.listenerList.remove(SourceStateChangedListener.class, listener);
    }
    
    private void fireSourceStateChanged(SourceStateChangedEvent event)
    {
    	Object[] listeners = listenerList.getListenerList();
    	
    	// Each listener occupies two elements - the first is the listener class and the second is the listener instance
    	for (int i = 0; i < listeners.length; i += 2)
    	{
    		if (listeners[i] == SourceStateChangedListener.class)
    		{
    			((SourceStateChangedListener)listeners[i + 1]).sourceStateChanged(event);
    		}
    	}
    }
    
    /**
     * Listens for changes to the source state and raises events.
     */
    private void sourceStateListen()
    {
        try
        {
            ALSourceState lastSourceState = ALSourceState.INITIAL;
            while (true)
            {
                if (lastSourceState != this.getSourceState())
                {
                    lastSourceState = this.getSourceState();
                    this.fireSourceStateChanged(new SourceStateChangedEvent(this));
                }

                Thread.sleep(1);
            }
        }
        catch (InterruptedException e)
        {
        }
    }
}

File: cs185/StudentSelecter/CS/StudentSelecter/WavePlayer.cs
namespace StudentSelecter
{
    using System;
    using System.Diagnostics;
    using System.IO;
    using System.Threading;
    using OpenTK.Audio;
    using OpenTK.Audio.OpenAL;

    /// <summary>
    /// Represents a wave file player.
    /// </summary>
    public sealed class WavePlayer
    {
        /// <summary>
        /// The OpenAL audio content used for playback.
        /// </summary>
        private AudioContext context;

        /// <summary>
        /// The OpenAL sound buffer identifier.
        /// </summary>
        private int buffer;

        /// <summary>
        /// The OpenAL sound source identifier.
        /// </summary>
        private int source;

        /// <summary>
        /// The source state listener thread.
        /// </summary>
        private Thread thread;

        /// <summary>
        /// Initializes a new instance of the <see cref="WavePlayer"/> class.
        /// </summary>
        /// <param name="fileName">The file to read audio data from.</param>
        public WavePlayer(string fileName)
            : this()
        {
            this.Load(fileName);
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="WavePlayer"/> class.
        /// </summary>
        /// <param name="stream">The stream to read audio data from.</param>
        public WavePlayer(Stream stream)
            : this()
        {
            this.Load(stream);
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="WavePlayer"/> class.
        /// </summary>
        private WavePlayer()
        {
            this.context = new AudioContext();

            this.buffer = AL.GenBuffer();
            this.source = AL.GenSource();

            this.thread = new Thread(new ThreadStart(this.SourceStateListen));
            this.thread.Start();
        }

        /// <summary>
        /// Raised when the <see cref="SourceState"/> property changes.
        /// </summary>
        public event EventHandler SourceStateChanged = delegate { };

        /// <summary>
        /// Gets the state of the audio source; whether it is playing, paused or stopped.
        /// </summary>
        public ALSourceState SourceState
        {
            get
            {
                int state;
                AL.GetSource(this.source, ALGetSourcei.SourceState, out state);
                return (ALSourceState)state;
            }
        }

        /// <summary>
        /// Loads a wave/riff audio file.
        /// </summary>
        /// <param name="stream">The stream to read from.</param>
        /// <param name="channels">The number of channels in the wave file.</param>
        /// <param name="bits">The number of bits per sample in the wave file.</param>
        /// <param name="rate">The sample rate of the wave file.</param>
        /// <returns>The sound data of the wave file.</returns>
        public static byte[] LoadWave(Stream stream, out int channels, out int bits, out int rate)
        {
            if (stream == null)
            {
                throw new ArgumentNullException("stream");
            }

            using (BinaryReader reader = new BinaryReader(stream))
            {
                // RIFF header signature
                if (new string(reader.ReadChars(4)) != "RIFF")
                {
                    throw new NotSupportedException("Specified stream is not a wave file.");
                }

                int riff_chunck_size = reader.ReadInt32();

                if (new string(reader.ReadChars(4)) != "WAVE")
                {
                    throw new NotSupportedException("Specified stream is not a wave file.");
                }

                // WAVE header format signature
                if (new string(reader.ReadChars(4)) != "fmt ")
                {
                    throw new NotSupportedException("Specified wave file is not supported.");
                }

                int format_chunk_size = reader.ReadInt32();
                int audio_format = reader.ReadInt16();
                int num_channels = reader.ReadInt16();
                int sample_rate = reader.ReadInt32();
                int byte_rate = reader.ReadInt32();
                int block_align = reader.ReadInt16();
                int bits_per_sample = reader.ReadInt16();

                string data_signature = new string(reader.ReadChars(4));
                if (data_signature != "data")
                {
                    throw new NotSupportedException("Specified wave file is not supported.");
                }

                int data_chunk_size = reader.ReadInt32();

                channels = num_channels;
                bits = bits_per_sample;
                rate = sample_rate;

                return reader.ReadBytes((int)reader.BaseStream.Length - (int)reader.BaseStream.Position);
            }
        }

        /// <summary>
        /// Gets the OpenAL format specifier for the sound format with the specified number of channels and bits.
        /// </summary>
        /// <param name="channels">The number of channels in the sound format.</param>
        /// <param name="bits">The number of bits in the sound format.</param>
        /// <returns>See summary.</returns>
        public static ALFormat GetSoundFormat(int channels, int bits)
        {
            switch (channels)
            {
                case 1:
                    return bits == 8 ? ALFormat.Mono8 : ALFormat.Mono16;
                case 2:
                    return bits == 8 ? ALFormat.Stereo8 : ALFormat.Stereo16;
                default:
                    throw new NotSupportedException("The specified sound format is not supported.");
            }
        }

        /// <summary>
        /// Loads a wave file into the player.
        /// </summary>
        /// <param name="fileName">The name of the file to load.</param>
        public void Load(string fileName)
        {
            this.Load(File.Open(fileName, FileMode.Open));
        }

        /// <summary>
        /// Loads a wave file into the player.
        /// </summary>
        /// <param name="stream">The stream to load from.</param>
        public void Load(Stream stream)
        {
            int channels, bits_per_sample, sample_rate;
            byte[] sound_data = WavePlayer.LoadWave(stream, out channels, out bits_per_sample, out sample_rate);
            AL.BufferData(this.buffer, WavePlayer.GetSoundFormat(channels, bits_per_sample), sound_data, sound_data.Length, sample_rate);

            AL.Source(this.source, ALSourcei.Buffer, this.buffer);
        }

        /// <summary>
        /// Plays the loaded wave file.
        /// </summary>
        public void Play()
        {
            AL.SourcePlay(this.source);
        }

        /// <summary>
        /// Frees resources.
        /// </summary>
        public void Dispose()
        {
            if (this.thread != null && this.thread.IsAlive)
            {
                this.thread.Abort();
            }

            AL.SourceStop(this.source);
            AL.DeleteSource(this.source);
            AL.DeleteBuffer(this.buffer);
            this.context.Dispose();
        }

        /// <summary>
        /// Listens for changes to the source state and raises events.
        /// </summary>
        private void SourceStateListen()
        {
            try
            {
                ALSourceState lastSourceState = ALSourceState.Initial;
                while (true)
                {
                    if (lastSourceState != this.SourceState)
                    {
                        lastSourceState = this.SourceState;
                        this.SourceStateChanged(this, EventArgs.Empty);
                    }

                    Thread.Sleep(1);
                }
            }
            catch (ThreadAbortException)
            {
            }
        }
    }
}
